### PR TITLE
Simplify `DefaultConnectionPool.OpenConcurrencyLimiter.tryLock`

### PR DESCRIPTION
**Pull request by stIncMale**
TODO: rebase on top of `master` after merging https://github.com/mongodb/mongo-java-driver/pull/699, then request for review.

JAVA-3927

### CODE REVIEW

File: driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java [hunk 1]
```diff
@@ -838,7 +838,7 @@ void openAsyncOrGetAvailable(
         private PooledConnection acquirePermitOrGetAvailableOpenedConnection(final boolean tryGetAvailable, final Timeout timeout)
                 throws MongoTimeoutException {
             PooledConnection availableConnection = null;
-            tryLock(timeout);
+            lockInterruptibly(lock);
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java:841]: Why is it ok to ignore the timeout when acquiring `lock`?

**Review by jyemin — COMMENTED** [review]
Can you provide some context as to the purpose of this change?  It seems like a behavioral change, but the PR title seems to indicate that it's not.

File: driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java (hunk 1 above)
Comment (stIncMale) [reply driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java:841]: To some extent this is answered in the commit message (which I modified several times, which resulted in re-creating the only commit in this PR again and again). But since this is still not obvious, let's think about a situation when locking may take time significant enough for us to want to make the locking itself timed.

The time to acquire this fair reentrant lock is `numberOfContendingThreads * averageTimeBeforeCallingWait`. `numberOfContendingThreads` is the number of threads which are ahead of the considered thread in the imaginary queue of threads that concurrently want to acquire the lock; `averageTimeBeforeCallingWait` is usually the average time to check the condition, but in our case it's more than that.

`numberOfContendingThreads`
Most of the time this lock is uncontended because threads get opened connections, and only after invalidation of the pool (or after creation of the pool) the lock has a chance to be contended.

`averageTimeBeforeCallingWait`
Has a potential to take some time only if the pool has many stale connections, which may happen if it has recently been invalidated and has large `maxSize`. While any individual invocation of `getPooledConnectionImmediately` has a potential to be slightly time-consuming (and, therefore, occasionally acquiring the lock may take comparable time), on average this method is of constant small time because the majority of invocations observe no stale connections.

If we also take into account that
1. all timeout implementations always overtime the specified timeout and that
2. we have a much more significant way to overtime: all asynchronous check outs and all (internal to the pool) asynchronous open requests are queued (open requests are also throttled) because they are executed by a single thread

worrying about the time to acquire the lock here does not seem important enough anymore. These thoughts appeared to me only recently, and not when I was implementing the functionality.

Comment (jyemin) [reply driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java:841]: Thanks for the more detailed explanation.  And really, the problem was that I didn't even look at the commit message.  Usually with single commit PRs the commit message is copied to the PR description automatically.  Perhaps that didn't happen because, as you said, you changed the message a few times.

File: driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java [hunk 2]
```diff
@@ -945,26 +945,13 @@ void tryHandOverOrRelease(final UsageTrackingInternalConnection openConnection)
             }
         }
 
-        /**
-         * @param timeout If {@linkplain Timeout#isInfinite() infinite},
-         *                then the lock is {@linkplain Lock#lockInterruptibly() acquired interruptibly}.
-         */
-        private void tryLock(final Timeout timeout) throws MongoTimeoutException {
-            boolean success;
+        private void lockInterruptibly(final Lock lock) throws MongoTimeoutException {
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java:948]: Why the change to take the lock as the parameter?  Is there more than lock that we could be using in this context?
Taking the lock as a param it seems more like a utility method and therefore should be static. 

What do you think?

Comment (jyemin) [inline driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java:948]: You can remove the throws clause now.

**Review by jyemin — CHANGES_REQUESTED** [review]

File: driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java (hunk 2 above)
Comment (stIncMale) [reply driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java]: Thanks, it should have been `MongointerruptedException`. Done.

Comment (stIncMale) [reply driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java]: > Why the change to take the lock as the parameter? Is there more than lock that we could be using in this context?

The change is to remove the weird asymmetry. Previously the code looked like
```java
lockInterruptibly(); // use a method that uses the `lock` field internally
try {
  ...
} finally {
  lock.unlock(); // use the `lock` field explicitly
}
```
now it looks like
```java
lockInterruptibly(lock); // use the `lock` field explicitly
try {
  ...
} finally {
  lock.unlock(); // use the `lock` field explicitly
}
```

> a utility method and therefore should be static

I 100% agree, and I wish we could write static methods in inner classes, but Java syntax forbids that (this restriction seems to be artificial and not helpful).

**Comment by stIncMale** [comment]
> Can you provide some context as to the purpose of this change? It seems like a behavioral change, but the PR title seems to indicate that it's not.

@jyemin Now that you know there is a commit message, is this request still actual?

**Review by jyemin — APPROVED** [review]
LGTM!
